x/auth/vesting/types/vesting_account.go

```diff
@@ -611,3 +589,393 @@ func marshalYaml(i interface{}) (interface{}, error) {
+	}
+	if !originalVesting.IsEqual(tva.OriginalVesting) {
+		return errors.New("original vesting coins does not match the sum of all coins in vesting periods")
+	}
+
+	return tva.BaseVestingAccount.Validate()
+}
+
+func (pva TrueVestingAccount) String() string {
+	out, _ := pva.MarshalYAML()
+	return out.(string)
+}
+
+// MarshalYAML returns the YAML representation of a TrueVestingAccount.
+func (pva TrueVestingAccount) MarshalYAML() (interface{}, error) {
+	accAddr, err := sdk.AccAddressFromBech32(pva.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	out := vestingAccountYAML{
+		Address:          accAddr,
+		AccountNumber:    pva.AccountNumber,
+		PubKey:           getPKString(pva),
+		Sequence:         pva.Sequence,
+		OriginalVesting:  pva.OriginalVesting,
+		DelegatedFree:    pva.DelegatedFree,
+		DelegatedVesting: pva.DelegatedVesting,
+		EndTime:          pva.EndTime,
+		StartTime:        pva.StartTime,
+		VestingPeriods:   pva.VestingPeriods,
+	}
+	return marshalYaml(out)
+}
+
+// ComputeClawback returns an account with all future vesting events removed,
+// plus the total sum of these events. When removing the future vesting events,
+// the lockup schedule will also have to be capped to keep the total sums the same.
+// (But future unlocking events might be preserved if they unlock currently vested coins.)
+// If the amount returned is zero, then the returned account should be unchanged.
+// Does not adjust DelegatedVesting
+func (tva TrueVestingAccount) ComputeClawback(clawbackTime int64) (TrueVestingAccount, sdk.Coins) {
+	// Compute the truncated vesting schedule and amounts.
+	// Work with the schedule as the primary data and recompute derived fields, e.g. OriginalVesting.
+	t := tva.StartTime
+	totalVested := sdk.NewCoins()
+	totalUnvested := sdk.NewCoins()
+	unvestedIdx := 0
+	for i, period := range tva.VestingPeriods {
+		t += period.Length
+		// tie in time goes to clawback
+		if t < clawbackTime {
+			totalVested = totalVested.Add(period.Amount...)
+			unvestedIdx = i + 1
+		} else {
+			totalUnvested = totalUnvested.Add(period.Amount...)
+		}
+	}
+	newVestingPeriods := tva.VestingPeriods[:unvestedIdx]
+
+	// To cap the unlocking schedule to the new total vested, conjunct with a limiting schedule
+	capPeriods := []Period{
+		{
+			Length: 0,
+			Amount: totalVested,
+		},
+	}
+	_, _, newLockupPeriods := ConjunctPeriods(tva.StartTime, tva.StartTime, tva.LockupPeriods, capPeriods)
+
+	_, _, newCombinedPeriods := ConjunctPeriods(tva.StartTime, tva.StartTime, newLockupPeriods, newVestingPeriods)
+
+	// Now construct the new account state
+	tva.OriginalVesting = totalVested
+	tva.EndTime = t
+	tva.LockupPeriods = newLockupPeriods
+	tva.VestingPeriods = newVestingPeriods
+	tva.CombinedPeriods = newCombinedPeriods
+	// DelegatedVesting will be adjusted elsewhere
+
+	return tva, totalUnvested
+}
+
+// Clawback transfers unvested tokens in a TrueVestingAccount to dest.
+// Future vesting events are removed. Unstaked tokens
```

Comment:
Incomplete thought?

Reply:
I'm lucky to have so many complete thoughts recorded. Fixed.